lib/puppet/provider/zabbix_proxy/ruby.rb

```diff
@@ -33,39 +23,178 @@ def self.prefetch(resources)
+    )
+  end
+
+  # Convert from proxy hash returned by API into a resource properties hash
+  def self.proxy_properties_hash(p)
+    {
+      ensure: :present,
+      # Proxy object properties
+      proxyid: p['proxyid'].to_i,
+      name:    p['host'],
+      mode:    status_to_mode(p['status']),
+      # Proxy Interface object properties
+      interfaceid: p['interface'].is_a?(Hash) ? p['interface']['interfaceid'] : nil,
+      ipaddress:   p['interface'].is_a?(Hash) ? p['interface']['ip'] : nil,
+      use_ip:      if p['interface'].is_a?(Hash) then p['interface']['useip'] == '1' ? :true : :false end,
+      port:        p['interface'].is_a?(Hash) ? p['interface']['port'].to_i : nil
+    }
+  end
+
   def create
-    # Set some vars
-    host = @resource[:hostname]
-    ipaddress = @resource[:ipaddress]
-
-    # Normally 0 is active and 1 is passive, in the API, its 5 and 6
-    proxy_mode = @resource[:mode] + 5
-
-    use_ip = @resource[:use_ip]
-    port = @resource[:port]
-
-    # Check if we need to connect via ip or fqdn
-    use_ip = use_ip ? 1 : 0
-
-    zbx.proxies.create_or_update(
-      host: host,
-      status: proxy_mode,
-      interfaces: [
-        ip: ipaddress,
-        dns: host,
-        useip: use_ip,
-        port: port
-      ]
+    mode = @resource[:mode] || :active
+
+    if mode == :passive
+      useip = if @resource[:use_ip].nil?
+                1
+              else
+                @resource[:use_ip] == :true ? 1 : 0
+              end
+      interface = {
+        ip: @resource[:ipaddress] || '127.0.0.1',
+        dns: @resource[:hostname],
+        useip: useip,
+        port: @resource[:port] || 10_051
+      }
+    else
+      interface = nil
+    end
+
+    zbx.proxies.create(
+      {
+        host: @resource[:hostname],
+        status: self.class.mode_to_status(mode),
+        interface: interface
+      }.compact
     )
+    @property_flush[:created] = true
   end
 
   def exists?
-    check_proxy(@resource[:hostname])
+    @property_hash[:ensure] == :present
   end
 
   def destroy
     zbx.proxies.delete([zbx.proxies.get_id(host: @resource[:hostname])].flatten)
+    @property_flush[:destroyed] = true
   end
 
-  mk_resource_methods
+  def flush
+    update unless @property_flush[:created] || @property_flush[:destroyed]
+
+    # Update @property_hash so that the output of puppet resource is correct
+    if @property_flush[:destroyed]
+      @property_hash.clear
+      @property_hash[:ensure] = :absent
+    else
+      proxy = self.class.proxies(@property_hash[:proxyid]).find { |p| p['host'] == @resource[:hostname] }
+      @property_hash = self.class.proxy_properties_hash(proxy)
+    end
+  end
+
+  def mode=(value)
+    @property_flush[:mode] = value
+  end
+
+  def change_to_passive_proxy
+    Puppet.debug("We're changing to a passive proxy")
+    # We need to call zbx.proxies.update with an `interface` hash.
+
+    useip = if @resource[:use_ip].nil?
+              1 # Default to true
```

Comment:
Should we just default this on the at the type?

Reply:
No, otherwise the property is always managed.

Reply:
> The defaultto method is used to specify default values for any attribute. Defaults are typically only specified for parameters. They are less common for properties because a default value will result in that property always being managed (there is no way to unset a value for the property).

From https://www.amazon.com/Puppet-Types-Providers-Extending-Ruby/dp/1449339328

Reply:
Interesting 🤔